src/server/ua_subscription_events.c

```diff
@@ -409,6 +409,593 @@ orOperator(UA_Server *server, UA_Session *session,
+        rule = 2;
+    } else if(castRule == 0 &&
+              (firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_STRING ||
+               firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BYTESTRING)){
+        rule = 3;
+    } else if(castRule == 0 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_NODEID){
+        rule = 4;
+    } else if(castRule == 0 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BOOLEAN){
+        rule = 5;
+    } else if(castRule == 0 &&
+              (firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_DOUBLE ||
+               firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_FLOAT)){
+        rule = 6;
+    } else if(castRule == 0 &&
+             firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_EXPANDEDNODEID){
+        rule = 7;
+    } else if(castRule == 0 &&
+             firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_GUID){
+        rule = 8;
+    } else if(castRule == 0 &&
+             firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_STATUSCODE){
+        rule = 9;
+    } else if(castRule == 0 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_LOCALIZEDTEXT){
+        rule = 10;
+    } else if(castRule == 0 &&
+             firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_QUALIFIEDNAME){
+        rule = 11;
+    } else if(castRule == 0 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_XMLELEMENT){
+        rule = 12;
+    } else if(castRule == 1 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BOOLEAN &&
+              isNumericUnsigned(secondCompareOperand->type->typeKind)){
+        rule = 100;
+    } else if(castRule == 1 &&
+              firstCompareOperand->type->typeKind == (UA_UInt32) UA_DATATYPEKIND_BOOLEAN &&
+              isNumericSigned(secondCompareOperand->type->typeKind)) {
+        rule = 101;
+    } else if(castRule == 1){
+        //todo error handling
+    }
+
+    switch(op) {
+    case UA_FILTEROPERATOR_EQUALS:
+        switch(rule) {
+        case 1:{
+            UA_UInt64 uint64_o1 = 0, uint64_o2 = 0;
+            implicitNumericCast(firstCompareOperand, &UA_TYPES[UA_TYPES_UINT64], &uint64_o1);
+            implicitNumericCast(secondCompareOperand, &UA_TYPES[UA_TYPES_UINT64], &uint64_o2);
+            if(uint64_o1 == uint64_o2)
+                return UA_STATUSCODE_GOOD;
+            break;}
+        case 2:{
+            UA_Int64 int64_o1 = 0, int64_o2 = 0;
+            implicitNumericCast(firstCompareOperand, &UA_TYPES[UA_TYPES_INT64], &int64_o1);
+            implicitNumericCast(secondCompareOperand, &UA_TYPES[UA_TYPES_INT64], &int64_o2);
+            if(int64_o1 == int64_o2)
+                return UA_STATUSCODE_GOOD;
+            break;}
+        case 3:
```

Comment:
Check the definition of `UA_order`.
With that you can cover most of these comparison operations with a single code-path.
Furthermore we can also check the equality of structure-typed values.